### PR TITLE
Migrating SmartStore tests to new testing framework

### PR DESCRIPTION
**Pull request by bhariharan**
I took the opportunity to rewrite the `Activity` based tests using `Espresso`, which is the new way. All tests are passing locally.

### CODE REVIEW

File: libs/SmartStore/build.gradle [hunk 1]
```diff
@@ -16,6 +16,9 @@ apply plugin: 'com.jfrog.bintray'
 dependencies {
     api project(':libs:SalesforceSDK')
     api 'net.zetetic:android-database-sqlcipher:3.5.7'
+    androidTestApi 'com.android.support.test:runner:1.0.1'
+    androidTestApi 'com.android.support.test:rules:1.0.1'
+    androidTestApi 'com.android.support.test.espresso:espresso-core:3.0.1'
```
Comment (bhariharan) [inline libs/SmartStore/build.gradle]: Include `espresso` for UI tests (`SmartStoreInspectorActivityTest` in this case).

File: libs/SmartStore/build.gradle [hunk 2]
```diff
@@ -48,13 +51,14 @@ android {
...
     testApplicationId "com.salesforce.androidsdk.smartstore.tests"
-    testInstrumentationRunner "com.salesforce.androidsdk.analytics.util.test.JUnitReportTestRunner"
+    testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
   }
   packagingOptions {
     exclude 'META-INF/LICENSE'
     exclude 'META-INF/LICENSE.txt'
     exclude 'META-INF/DEPENDENCIES'
     exclude 'META-INF/NOTICE'
+    pickFirst 'protobuf.meta'
```
Comment (bhariharan) [inline libs/SmartStore/build.gradle]: This `Gradle` compilation conflict is caused by `espresso`.

File: libs/test/SmartStoreTest/src/com/salesforce/androidsdk/store/SmartStoreInspectorActivityTest.java [hunk 3]
```diff
@@ -65,16 +79,13 @@
 	private Context targetContext;
 	private SmartStore store;
 
-	public SmartStoreInspectorActivityTest() {
-		super(SmartStoreInspectorActivity.class);
-	}
+	@Rule
+    public ActivityTestRule<SmartStoreInspectorActivity> smartStoreInspectorActivityTestRule = new ActivityTestRule<SmartStoreInspectorActivity>(SmartStoreInspectorActivity.class);
```
Comment (bhariharan) [inline libs/test/SmartStoreTest/src/com/salesforce/androidsdk/store/SmartStoreInspectorActivityTest.java]: Rule specifying which `Activity` to launch for this test.

File: libs/test/SmartStoreTest/src/com/salesforce/androidsdk/store/SmartStoreInspectorActivityTest.java [hunk 4]
```diff
@@ -190,15 +207,15 @@ public void testClickingRunWithValidQueryAndPageSizeAndPageIndex() {
 	/**
 	 * Testing autocomplete
 	 */
+    @Test
 	public void testAutoComplete() {
-		MultiAutoCompleteTextView queryTextView = (MultiAutoCompleteTextView) getActivity()
+		MultiAutoCompleteTextView queryTextView = smartStoreInspectorActivityTestRule.getActivity()
```
Comment (bhariharan) [inline libs/test/SmartStoreTest/src/com/salesforce/androidsdk/store/SmartStoreInspectorActivityTest.java]: The rule can be used to access the `Activity` instance itself.

File: libs/test/SmartStoreTest/src/com/salesforce/androidsdk/store/SmartStoreInspectorActivityTest.java [hunk 5]
```diff
@@ -281,59 +282,31 @@ private void checkInspectorState(String query, String pageSize,
...
         try {
-            runTestOnUiThread(new Runnable() {
-                @Override public void run() {
-                    TextView v = (TextView) getActivity().findViewById(textViewId);
-                    v.setText(text);
-                    if (v instanceof EditText)
-                        ((EditText) v).setSelection(v.getText().length());
-                }
-            });
+            onView(withId(textViewId)).perform(replaceText(text), closeSoftKeyboard());
```
Comment (bhariharan) [inline libs/test/SmartStoreTest/src/com/salesforce/androidsdk/store/SmartStoreInspectorActivityTest.java]: Much easier to insert text or perform clicks and other UI actions with views using `Espresso`.

**Comment by brandonpage** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Tests have failed, see below for more information.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR, try to keep changes smaller if you can</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Needs testing on a Phone if change is non-trivial</td>
    </tr>
  </tbody>
</table>

### Tests: 

Name | Classname | Time|
--- | --- | ---|
testGetInfo | com.salesforce.androidsdk.phonegap.SDKInfoJSTest | 5.943|
testGetSDKInfo | com.salesforce.androidsdk.phonegap.SDKInfoPluginTest | 0.404|
testUpsertConcurrentEntries | com.salesforce.androidsdk.phonegap.SmartStoreLoadJSTest | 1.108|
testRefreshSyncDown | com.salesforce.androidsdk.phonegap.SmartSyncJSTest | 0.454|
testSyncSObjectDetectConflictUpdate | com.salesforce.androidsdk.phonegap.SmartSyncJSTest | 0.378|
testSyncUpLocallyDeleted | com.salesforce.androidsdk.phonegap.SmartSyncJSTest | 0.353|
testSyncUpLocallyUpdatedWithNoOverwrite | com.salesforce.androidsdk.phonegap.SmartSyncJSTest | 0.328|
No tests found. | com.android.builder.testing.ConnectedDevice | 0.001|


<p align="right" data-meta="generated_by_ci/circleci">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by bhariharan** [comment]
It has been more than a year since we moved the templates to their own repo, but we left the `TemplateAppTests` around in an orphaned state. I moved the 2 tests in there, `ServerPickerActivityTest` and `PasscodeActivityTest` into `RestExplorerTests` and rewrote them using `Espresso`. They can now be run. I also removed the defunct `TemplateAppTests` shell and updated some rogue references in `README` and legacy `build` scripts.

**Comment by bhariharan** [comment]
Merging this to get `CircleCI` going with the new tests. Next PR will have `SmartSync` and `SalesforceHybrid` converted.

**Pull request merged**